File: uploadservice/src/main/java/net/gotev/uploadservice/MultipartUploadTask.java
package net.gotev.uploadservice;

import net.gotev.uploadservice.data.NameValue;
import net.gotev.uploadservice.data.UploadFile;
import net.gotev.uploadservice.network.BodyWriter;

import java.io.IOException;
import java.nio.charset.Charset;

/**
 * Implements an HTTP Multipart upload task.
 *
 * @author gotev (Aleksandar Gotev)
 * @author eliasnaur
 * @author cankov
 */
public class MultipartUploadTask extends HttpUploadTask {

    private static final String BOUNDARY_SIGNATURE = "-------AndroidUploadService";
    private static final Charset US_ASCII = Charset.forName("US-ASCII");
    private static final String NEW_LINE = "\r\n";
    private static final String TWO_HYPHENS = "--";

    // properties associated to each file
    protected static final String PROPERTY_REMOTE_FILE_NAME = "httpRemoteFileName";
    protected static final String PROPERTY_CONTENT_TYPE = "httpContentType";
    protected static final String PROPERTY_PARAM_NAME = "httpParamName";

    private byte[] boundaryBytes;
    private byte[] trailerBytes;
    private Charset charset = Charset.forName("UTF-8");

    @Override
    public void performInitialization() {

        String boundary = BOUNDARY_SIGNATURE + System.nanoTime();
        boundaryBytes = (TWO_HYPHENS + boundary + NEW_LINE).getBytes(US_ASCII);
        trailerBytes = (TWO_HYPHENS + boundary + TWO_HYPHENS + NEW_LINE).getBytes(US_ASCII);

        HttpUploadTaskParameters httpParams = getHttpParams();

        if (params.getFiles().size() <= 1) {
            httpParams.addHeader("Connection", "close");
        } else {
            httpParams.addHeader("Connection", "Keep-Alive");
        }

        httpParams.addHeader("Content-Type", "multipart/form-data; boundary=" + boundary);
    }

    @Override
    public long getBodyLength() {
        return (getRequestParametersLength() + getFilesLength() + trailerBytes.length);
    }

    @Override
    public void onWriteRequestBody(BodyWriter bodyWriter) throws IOException {
        //reset uploaded bytes when the body is ready to be written
        //because sometimes this gets invoked when network changes
        resetUploadedBytes();
        writeRequestParameters(bodyWriter);
        writeFiles(bodyWriter);
        bodyWriter.write(trailerBytes);
        broadcastProgress(trailerBytes.length);
    }

    private long getFilesLength() {
        long total = 0;

        for (UploadFile file : params.getFiles()) {
            total += getTotalMultipartBytes(file);
        }

        return total;
    }

    private long getRequestParametersLength() {
        long parametersBytes = 0;
        HttpUploadTaskParameters params = getHttpParams();

        if (!params.getRequestParameters().isEmpty()) {
            for (final NameValue parameter : params.getRequestParameters()) {
                // the bytes needed for every parameter are the sum of the boundary bytes
                // and the bytes occupied by the parameter
                parametersBytes += boundaryBytes.length + getMultipartBytes(parameter).length;
            }
        }

        return parametersBytes;
    }

    private byte[] getMultipartBytes(NameValue parameter) {
        return ("Content-Disposition: form-data; name=\"" + parameter.getName() + "\""
                + NEW_LINE + NEW_LINE + parameter.getValue() + NEW_LINE).getBytes(charset);
    }

    private byte[] getMultipartHeader(UploadFile file) {
        String header = "Content-Disposition: form-data; name=\"" +
                file.getProperties().get(PROPERTY_PARAM_NAME) + "\"; filename=\"" +
                file.getProperties().get(PROPERTY_REMOTE_FILE_NAME) + "\"" + NEW_LINE +
                "Content-Type: " + file.getProperties().get(PROPERTY_CONTENT_TYPE) +
                NEW_LINE + NEW_LINE;

        return header.getBytes(charset);
    }

    private long getTotalMultipartBytes(UploadFile file) {
        return boundaryBytes.length + getMultipartHeader(file).length + file.getHandler().size(context)
                + NEW_LINE.getBytes(charset).length;
    }

    private void writeRequestParameters(BodyWriter bodyWriter) throws IOException {
        HttpUploadTaskParameters params = getHttpParams();

        if (!params.getRequestParameters().isEmpty()) {
            for (final NameValue parameter : params.getRequestParameters()) {
                bodyWriter.write(boundaryBytes);
                byte[] formItemBytes = getMultipartBytes(parameter);
                bodyWriter.write(formItemBytes);

                broadcastProgress(boundaryBytes.length + formItemBytes.length);
            }
        }
    }

    private void writeFiles(BodyWriter bodyWriter) throws IOException {
        for (UploadFile file : params.getFiles()) {
            if (!getShouldContinue())
                break;

            bodyWriter.write(boundaryBytes);
            byte[] headerBytes = getMultipartHeader(file);
            bodyWriter.write(headerBytes);

            broadcastProgress(boundaryBytes.length + headerBytes.length);

            bodyWriter.writeStream(file.getHandler().stream(context), this);

            byte[] newLineBytes = NEW_LINE.getBytes(charset);
            bodyWriter.write(newLineBytes);

            broadcastProgress(newLineBytes.length);
        }
    }

    @Override
    protected void onSuccessfulUpload() {
        addAllFilesToSuccessfullyUploadedFiles();
    }

}

File: uploadservice/src/main/java/net/gotev/uploadservice/MultipartUploadTask.kt
package net.gotev.uploadservice

import net.gotev.uploadservice.data.NameValue
import net.gotev.uploadservice.data.UploadFile
import net.gotev.uploadservice.network.BodyWriter

/**
 * Implements an HTTP Multipart upload task.
 *
 * @author gotev (Aleksandar Gotev)
 * @author eliasnaur
 * @author cankov
 */
class MultipartUploadTask : HttpUploadTask() {

    companion object {
        private const val BOUNDARY_SIGNATURE = "-------AndroidUploadService"
        private const val NEW_LINE = "\r\n"
        private const val TWO_HYPHENS = "--"
        private val US_ASCII = Charsets.US_ASCII

        // properties associated to each file
        const val PROPERTY_REMOTE_FILE_NAME = "httpRemoteFileName"
        const val PROPERTY_CONTENT_TYPE = "httpContentType"
        const val PROPERTY_PARAM_NAME = "httpParamName"
    }

    private val charset = Charsets.UTF_8

    private val boundary by lazy {
        BOUNDARY_SIGNATURE + System.nanoTime()
    }

    private val boundaryBytes by lazy {
        (TWO_HYPHENS + boundary + NEW_LINE).toByteArray(US_ASCII)
    }

    private val trailerBytes by lazy {
        (TWO_HYPHENS + boundary + TWO_HYPHENS + NEW_LINE).toByteArray(US_ASCII)
    }

    private val newLineBytes by lazy {
        NEW_LINE.toByteArray(charset)
    }

    override val bodyLength: Long
        get() = requestParametersLength + filesLength + trailerBytes.size

    private val filesLength: Long
        get() = params.files.map { it.getTotalMultipartBytes() }.sum()

    // the bytes needed for every parameter are the sum of the boundary bytes
    // and the bytes occupied by the parameter
    private val requestParametersLength: Long
        get() {
            var parametersBytes: Long = 0
            val params = httpParams

            if (params.requestParameters.isNotEmpty()) {
                for (parameter in params.requestParameters) {
                    parametersBytes += (boundaryBytes.size + parameter.getMultipartBytes().size).toLong()
                }
            }

            return parametersBytes
        }

    private fun NameValue.getMultipartBytes(): ByteArray {
        return ("Content-Disposition: form-data; name=\"" + name + "\""
                + NEW_LINE + NEW_LINE + value + NEW_LINE).toByteArray(charset)
    }

    private fun UploadFile.getMultipartHeader(): ByteArray {
        val header = "Content-Disposition: form-data; name=\"" +
                properties[PROPERTY_PARAM_NAME] + "\"; filename=\"" +
                properties[PROPERTY_REMOTE_FILE_NAME] + "\"" + NEW_LINE +
                "Content-Type: " + properties[PROPERTY_CONTENT_TYPE] +
                NEW_LINE + NEW_LINE

        return header.toByteArray(charset)
    }

    private fun UploadFile.getTotalMultipartBytes(): Long {
        return boundaryBytes.size.toLong() +
                getMultipartHeader().size.toLong() +
                handler.size(context) +
                newLineBytes.size.toLong()
    }

    private fun BodyWriter.writeRequestParameters() {
        val params = httpParams

        if (params.requestParameters.isNotEmpty()) {
            for (parameter in params.requestParameters) {
                val formItemBytes = parameter.getMultipartBytes()
                write(boundaryBytes)
                write(formItemBytes)
                broadcastProgress((boundaryBytes.size + formItemBytes.size).toLong())
            }
        }
    }

    private fun BodyWriter.writeFiles() {
        for (file in params.files) {
            if (!shouldContinue) break

            val headerBytes = file.getMultipartHeader()

            write(boundaryBytes)
            write(headerBytes)
            broadcastProgress((boundaryBytes.size + headerBytes.size).toLong())

            writeStream(file.handler.stream(context), this@MultipartUploadTask)
            write(newLineBytes)
            broadcastProgress(newLineBytes.size.toLong())
        }
    }

    override fun performInitialization() {
        httpParams.apply {
            addHeader("Content-Type", "multipart/form-data; boundary=$boundary")
            addHeader("Connection", if (params.files.size <= 1) "close" else "Keep-Alive")
        }
    }

    override fun onWriteRequestBody(bodyWriter: BodyWriter) {
        //reset uploaded bytes when the body is ready to be written
        //because sometimes this gets invoked when network changes
        resetUploadedBytes()

        bodyWriter.apply {
            writeRequestParameters()
            writeFiles()
            write(trailerBytes)
        }

        broadcastProgress(trailerBytes.size.toLong())
    }

    override fun onSuccessfulUpload() {
        addAllFilesToSuccessfullyUploadedFiles()
    }

}
